File: ProyectoJonathan/Java/ejercicio4.java
package ejercicios;

import java.util.Scanner;

/*
Dada una matriz m × n, realice un algoritmo para voltearla de arriba hacia abajo a
derecha.
     Entrada:
       1  2  3
       4  5  6
       7  8  9
       10 11 12
     Salida: La matriz volteada de abajo a arriba es
       10 11 12
       7  8  9
       4  5  6
       1  2  3
 */

public class ejercicio4 {

    public static void main(String[] args) {
        Scanner leer = new Scanner(System.in);
        int n, m;
        System.out.println("Ingrese la cantidad de filas de la matriz: ");
        n = leer.nextInt();
        System.out.println("ngrese la cantidad de columnas de la matriz: ");
        m = leer.nextInt();

        int[][] Arreglo = new int[n][m];

        //Se leen los valores del arreglo
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                Arreglo[i][j] = leer.nextInt();
            }
        }
        
        //Se llama a la funcón que "ordena" el vector
        Arreglo = OrdenarVerticalmente(Arreglo, n, m);
        
        
        //Se imprime nuevamente el vector
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                System.out.print(Arreglo[i][j] + " ");
            }
            System.out.println("");
        }
    }

    public static int[][] OrdenarVerticalmente(int[][] Arreglo, int n, int m) {

        //Se mueve el vector de abajo hacía arriba, fila por fila, y 
        //se aumenta 1 en columnas cada que se recorre una fila
        int keeper = 0;
        for (int i = 0; i < m; i++) {
            for (int j = n - 1; j > 0; j--) {
                for (int k = 0; k < j; k++) {
                    keeper = Arreglo[j][i];
                    Arreglo[j][i] = Arreglo[k][i];
                    Arreglo[k][i] = keeper;
                    j--;
                }

            }

        }

        return Arreglo;
    }
}

File: ProyectoJonathan/Python/ejercicio4.py
import numpy

''' 
    Dada una matriz m × n, realice un algoritmo para voltearla de abajo hacia arriba.
     Entrada:
       1  2  3
       4  5  6
       7  8  9
       10 11 12
     Salida: La matriz volteada de abajo a arriba es
       10 11 12
       7  8  9
       4  5  6
       1  2  3
'''

def OrdenarVerticalmente(Arreglo,n, m):
    #Se mueve el vector de abajo hacía arriba, fila por fila, y se aumenta 1 en columnas cada que se recorre una fila
    
    for i in range(0, m):
        for j in range(n - 1, 0, -1):
            for k in range(0, j):
                keeper = Arreglo[j][i]
                Arreglo[j][i] = Arreglo[k][i]
                Arreglo[k][i] = keeper
                j-=1
    return Arreglo

n = int(input("Ingrese la cantidad de filas de la matriz: "))
m = int(input("Ingrese la cantidad de columnas de la matriz: "))

#Usamos el método "zeros" de numpy para generar una matriz "encerada", se especifica la cantidad de filas y columnas, además del tipo 
#de datos que contiene la matriz.

matriz = numpy.zeros([n,m] , dtype = int)
for i in range(0, n):
    for j in range(0, m):
        matriz[i][j] = int(input())
        
print("Matriz Original: ")
print(matriz)

matriz = OrdenarVerticalmente(matriz,int(n), int(m))

print("Matriz Volteada: ")
print(matriz)
